Allow re-use of deprecated permissions MODPERMS-130

adamdickmeiss opened the pull request:
This PR fixes the problems as described in MODPERMS-130.

It also fixes a problem where permission could get a new module context if the structure was the same. That means multiple modules could be enabled and definitions could start to differ for same permission.

craigmcnally commented on src/main/java/org/folio/rest/impl/TenantPermsAPI.java:
```diff
@@ -106,8 +106,8 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
-                    if (dbPerm != null && Boolean.FALSE.equals(dbPerm.getMutable())
-                        && PermissionUtils.equals(perm, dbPerm)) {
+                    if (Boolean.TRUE.equals(dbPerm.getDeprecated()) || (Boolean.FALSE.equals(dbPerm.getMutable())
+                        && PermissionUtils.equals(perm, dbPerm))) {
```
Can you please update the comment below?

adamdickmeiss replied:
Sure.

hjiebsco commented on src/main/java/org/folio/rest/impl/TenantPermsAPI.java:
```diff
@@ -106,8 +106,8 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
                     }
-                    if (dbPerm != null && Boolean.FALSE.equals(dbPerm.getMutable())
-                        && PermissionUtils.equals(perm, dbPerm)) {
+                    if (Boolean.TRUE.equals(dbPerm.getDeprecated()) || (Boolean.FALSE.equals(dbPerm.getMutable())
+                        && PermissionUtils.equals(perm, dbPerm))) {
```
Note `PermissionUtils.equals(perm, dbPerm)` will compare perm display name as well. What's the chance that the new perm will also have `(Deprecated)` as prefix?

skoczko replied:
@adamdickmeiss @craigmcnally Guys, shouldn't all comparison be done with the permission name (not displayName)?

adamdickmeiss replied:
@hjiebsco .. Doesn't matter.. It will fail regardless of display name and my test shows it.  This condition should *not* use PermissionUtils.equals as this is not relevant. Either there's a colliision because a permission is defined elsewhere or NOT.

It will be utterly confusing if we allowed same name in two modules and .. suddenly because you change the definition in one of the modules there *might* be a conflict. I say *might* because it might be depending on the order of insertion because the context only allows ONE module to be attached to a permission.

adamdickmeiss replied:
The permissionName is already the same.. The check can be simplified to:
```java
  if (Boolean.TRUE.equals(dbPerm.getDeprecated()) || Boolean.FALSE.equals(dbPerm.getMutable())) {
```

hjiebsco replied:
I will mark this as resolved. But we still need to decide the implication of re-use to downgrade.

hjiebsco reviewed (commented):
Would module downgrade still be supported if the context is overwritten by another module during reuse?

hjiebsco reviewed (commented):
If all modules define permission name based on certain naming convention, maybe we can avoid naming conflict / reuse between modules?

adamdickmeiss commented:
> Would module downgrade still be supported if the context is overwritten by another module during reuse?

Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.

adamdickmeiss commented:
> If all modules define permission name based on certain naming convention, maybe we can avoid naming conflict / reuse between modules?

Yes. It would have been nice to avoid these duplicates with some scheme or automatic way. Like interfaces which are also names, but there has been a restriction on that from day one.

hjiebsco commented:
> > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> 
> Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.

Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.

adamdickmeiss commented:
> > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > 
> > 
> > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> 
> Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
u



> > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > 
> > 
> > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> 
> Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.

userA will keep permissions I hope. So she should have permission to use modB , then later (back to) modA.

hjiebsco commented:
> > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > 
> > > 
> > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > 
> > 
> > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > u
> 
> > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > 
> > > 
> > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > 
> > 
> > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> 
> userA will keep permissions I hope. So she should have permission to use modB , then later (back to) modA.

But how can we know that userA should have the reused permA for modB?

adamdickmeiss commented:
> > > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > > 
> > > > 
> > > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > > 
> > > 
> > > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > > u
> > 
> > 
> > > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > > 
> > > > 
> > > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > > 
> > > 
> > > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > 
> > 
> > userA will keep permissions I hope. So she should have permission to use modB , then later (back to) modA.
> 
> But how can we know that userA should have the reused permA for modB?

When a permission is marked deprecated (and that happens when it's not part of a module definition anymore) and the context is correct.. Then if the system ALSO removed that permission for that user, then obviously that information is lost that a user had that permission.

Which is why I said I hope the user still has the permission for it (even though it's deprecated). That means, if another module (including the same module where it was originally defined) defines it again, then the user is not affected. And so downgrade would just work. I thought the idea of deprecated permissions was a soft delete. If at any time, two modules start to define the permission, then there should be a conflict.

Example.. A module defines a set of permissions.. But there's another module defiing the same permissionNames (say another implementation of mod-authtoken or mod-login-saml or something).. then we should be able to go from one to another.. One will be disabled and another one will be disabled.. It's not a something that is obscure .. It may happen.

hjiebsco commented:
> > > > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > > > 
> > > > > 
> > > > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > > > 
> > > > 
> > > > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > > > u
> > > 
> > > 
> > > > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > > > 
> > > > > 
> > > > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > > > 
> > > > 
> > > > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > > 
> > > 
> > > userA will keep permissions I hope. So she should have permission to use modB , then later (back to) modA.
> > 
> > 
> > But how can we know that userA should have the reused permA for modB?
> 
> When a permission is marked deprecated (and that happens when it's not part of a module definition anymore) and the context is correct.. Then if the system ALSO removed that permission for that user, then obviously that information is lost that a user had that permission.
> 
> Which is why I said I hope the user still has the permission for it (even though it's deprecated). That means, if another module (including the same module where it was originally defined) defines it again, then the user is not affected. And so downgrade would just work. I thought the idea of deprecated permissions was a soft delete. If at any time, two modules start to define the permission, then there should be a conflict.
> 
> Example.. A module defines a set of permissions.. But there's another module defiing the same permissionNames (say another implementation of mod-authtoken or mod-login-saml or something).. then we should be able to go from one to another.. One will be disabled and another one will be disabled.. It's not a something that is obscure .. It may happen.
> 
> I thought we'd have a special call or A UI to REALLY remove the deprecated permissions.. Perhaps it's done for all. Perhaps it's done individually.. But by default, it's a soft delete.
> 
> I probably misunderstood what meant @hjiebsco :-)

If the same permission is redefined by another implementation of the same interface, re-use makes sense because it is essentially the same permission for the same purpose. But if it is redefined by another unrelated module for a different purpose, the re-use will create a complicated scenario.

adamdickmeiss commented:
> > > > > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > > > > 
> > > > > > 
> > > > > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > > > > 
> > > > > 
> > > > > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > > > > u
> > > > 
> > > > 
> > > > > > > Would module downgrade still be supported if the context is overwritten by another module during reuse?
> > > > > > 
> > > > > > 
> > > > > > Upgrade/downgrade is not a concern here AFAIK. If a module transitions from one version to another the new version's permissions are posted.. Any permissions that is no longer a member of that module will be deprecated.. The context is allowed to be changed when it's deprecated.
> > > > > 
> > > > > 
> > > > > Let's use an example: modA has permA defined in v1 and userA has that permA assigned. Next, modA is upgraded to v2 and the permA in v1 is marked as deprecated but userA still has that permA assigned (soft delete as what has been decided and implemented). Then modB comes in and reused permA, what does that mean to the userA permA assignment? If you remove that assignment, what if modA needs to downgrade from v2 to v1? The userA will not have that permA assignment anymore.
> > > > 
> > > > 
> > > > userA will keep permissions I hope. So she should have permission to use modB , then later (back to) modA.
> > > 
> > > 
> > > But how can we know that userA should have the reused permA for modB?
> > 
> > 
> > When a permission is marked deprecated (and that happens when it's not part of a module definition anymore) and the context is correct.. Then if the system ALSO removed that permission for that user, then obviously that information is lost that a user had that permission.
> > Which is why I said I hope the user still has the permission for it (even though it's deprecated). That means, if another module (including the same module where it was originally defined) defines it again, then the user is not affected. And so downgrade would just work. I thought the idea of deprecated permissions was a soft delete. If at any time, two modules start to define the permission, then there should be a conflict.
> > Example.. A module defines a set of permissions.. But there's another module defiing the same permissionNames (say another implementation of mod-authtoken or mod-login-saml or something).. then we should be able to go from one to another.. One will be disabled and another one will be disabled.. It's not a something that is obscure .. It may happen.
> > I thought we'd have a special call or A UI to REALLY remove the deprecated permissions.. Perhaps it's done for all. Perhaps it's done individually.. But by default, it's a soft delete.
> > I probably misunderstood what meant @hjiebsco :-)
> 
> If the same permission is redefined by another implementation of the same interface, re-use makes sense because it is essentially the same permission for the same purpose. But if it is redefined by another unrelated module for a different purpose, the re-use will create a complicated scenario.

That could happen that somebody hijacked a permission. But so far this has never been the case.. When we have seen permissions being re-used they were for the same purpose. Again .. in order for this to happen one would have to bet on the other one being gone.

adamdickmeiss commented:
Closing as GitHub/GitHub seems to be stock.. Continued in https://github.com/folio-org/mod-permissions/pull/112
